File: ios/Sources/OfflineSpeechRecognitionPlugin/OfflineSpeechRecognitionPlugin.swift
import Capacitor
import AVFoundation
import Foundation

@objc(OfflineSpeechRecognitionPlugin)
public class OfflineSpeechRecognitionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let jsName = "OfflineSpeechRecognition"
    public let identifier = "OfflineSpeechRecognitionPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getSupportedLanguages", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDownloadedLanguageModels", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "downloadLanguageModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startRecognition", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopRecognition", returnType: CAPPluginReturnPromise)
    ]
    
    private var audioEngine: AVAudioEngine?
    private var processingQueue: DispatchQueue?
    private var voskModel: VoskModel?
    private var voskRecognizer: VoskRecognizer?
    private var isRecording = false
    private var currentLanguage = "en-us"
    private var modelDownloadManager: ModelDownloadManager?
    
    // supported languages mapping to model names - same as Android
    private let supportedLanguages: [String: String] = [
        "en-us": "model-en",
        "de": "model-de",
        "fr": "model-fr",
        "es": "model-es",
        "pt": "model-pt",
        "zh": "model-zh",
        "ru": "model-ru",
        "tr": "model-tr",
        "vi": "model-vi",
        "it": "model-it",
        "hi": "model-hi",
        "gu": "model-gu",
        "te": "model-te",
        "ja": "model-ja",
        "ko": "model-ko"
    ]
    
    // Language names mapping
    private let languageNames: [String: String] = [
        "hi": "Hindi",
        "ko": "Korean",
        "de": "German",
        "fr": "French",
        "te": "Telugu",
        "es": "Spanish",
        "zh": "Chinese",
        "ru": "Russian",
        "tr": "Turkish",
        "it": "Italian",
        "gu": "Gujarati",
        "ja": "Japanese",
        "pt": "Portuguese",
        "vi": "Vietnamese",
        "en-us": "English (US)",
        "en-in": "English (India)"
    ]
    
    override public func load() {
        super.load()
        print("SpeechToTextPlugin loaded")
        requestMicrophonePermission()
        modelDownloadManager = ModelDownloadManager()
        print("Model download manager initialized")
    }
    
    // MARK: - Plugin Methods
    
    @objc func getSupportedLanguages(_ call: CAPPluginCall) {
        var languages: [[String: Any]] = []
        
        for (code, modelName) in supportedLanguages {
            let language: [String: Any] = [
                "code": code,
                "modelName": modelName,
                "name": languageNames[code] ?? code,
                "offlineSupported": true
            ]
            languages.append(language)
        }
        
        call.resolve(["languages": languages])
    }
    
    @objc func getDownloadedLanguageModels(_ call: CAPPluginCall) {
        var models: [[String: Any]] = []
        
        guard let downloadManager = modelDownloadManager else {
            call.reject("Model download manager not initialized")
            return
        }
        
        for (code, modelName) in supportedLanguages {
            if downloadManager.isModelDownloaded(modelName: modelName) {
                let model: [String: Any] = [
                    "modelName": modelName,
                    "language": code,
                    "name": languageNames[code] ?? code,
                    "size": downloadManager.getModelSize(modelName: modelName),
                    "offlineSupported": true
                ]
                models.append(model)
            }
        }
        
        call.resolve(["models": models])
    }
    
    @objc func downloadLanguageModel(_ call: CAPPluginCall) {
        guard let language = call.getString("language") else {
            call.reject("Language parameter is required")
            return
        }
        
        guard let modelName = supportedLanguages[language] else {
            call.reject("Unsupported language: \(language)")
            return
        }
        
        guard let downloadManager = modelDownloadManager else {
            call.reject("Model download manager not initialized")
            return
        }
        
        // Check if model is already downloaded
        if downloadManager.isModelDownloaded(modelName: modelName) {
            call.resolve([
                "success": true,
                "language": language,
                "message": "Model already downloaded",
                "offlineSupported": true
            ])
            return
        }
        
        // Check if download is already in progress
        if downloadManager.isDownloadInProgress(modelName: modelName) {
            call.reject("Download already in progress for this model")
            return
        }
        
        // Start download
        downloadManager.downloadModel(
            modelName: modelName,
            onProgress: { progress in
                let progressData: [String: Any] = [
                    "progress": progress,
                    "message": "Downloading model... \(progress)%"
                ]
                self.notifyListeners("downloadProgress", data: progressData)
            },
            onSuccess: {
                call.resolve([
                    "success": true,
                    "language": language,
                    "message": "Model downloaded successfully",
                    "offlineSupported": true
                ])
            },
            onError: { error in
                call.reject("Error downloading model: \(error)")
            }
        )
    }
    
    @objc func startRecognition(_ call: CAPPluginCall) {
        let language = call.getString("language", "en-us")
        
        print("Starting recognition for language: \(language)")
        
        guard let modelName = supportedLanguages[language] else {
            print("ERROR: Unsupported language: \(language)")
            call.reject("Unsupported language: \(language)")
            return
        }
        
        guard let downloadManager = modelDownloadManager else {
            call.reject("Model download manager not initialized")
            return
        }
        
        // Check if model is downloaded
        if !downloadManager.isModelDownloaded(modelName: modelName) {
            call.reject("Language model not downloaded. Please download the model first.")
            return
        }
        
        // Check permissions
        guard checkMicrophonePermission() else {
            print("ERROR: Microphone permission not granted")
            call.reject("Microphone permission is required")
            return
        }
        
        do {
            try loadModel(modelName: modelName)
            startVoskRecognition(language: language)
            call.resolve()
        } catch {
            print("ERROR: Failed to load model: \(error)")
            call.reject("Failed to load model: \(error.localizedDescription)")
        }
    }
    
    @objc func stopRecognition(_ call: CAPPluginCall) {
        stopVoskRecognition()
        call.resolve()
    }
    
    // MARK: - Private Methods
    
    private func requestMicrophonePermission() {
        AVAudioSession.sharedInstance().requestRecordPermission { granted in
            DispatchQueue.main.async {
                if granted {
                    print("Microphone permission granted")
                } else {
                    print("Microphone permission denied")
                }
            }
        }
    }
    
    private func checkMicrophonePermission() -> Bool {
        let audioSession = AVAudioSession.sharedInstance()
        
        switch audioSession.recordPermission {
        case .granted:
            return true
        case .denied:
            return false
        case .undetermined:
            audioSession.requestRecordPermission { granted in
                // Permission result handled asynchronously
            }
            return false
        @unknown default:
            return false
        }
    }
    
    private func loadModel(modelName: String) throws {
        guard let downloadManager = modelDownloadManager else {
            throw NSError(domain: "SpeechToTextPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Model download manager not initialized"])
        }
        
        // Resolve the actual root dir (handles zips with a top-level folder)
        let modelDir = downloadManager.getResolvedModelDirectory(modelName: modelName)
        
        if !FileManager.default.fileExists(atPath: modelDir.path) {
            throw NSError(domain: "SpeechToTextPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Model directory not found: \(modelDir.path)"])
        }
        
        if !downloadManager.verifyModel(modelDir: modelDir) {
            throw NSError(domain: "SpeechToTextPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Model verification failed. Model files may be corrupted or incomplete."])
        }
        
        print("Loading Vosk model from: \(modelDir.path)")
        
        // Initialize processing queue if not already done
        if processingQueue == nil {
            processingQueue = DispatchQueue(label: "voskRecognizerQueue")
        }
        
        // Load the model
        voskModel = VoskModel(modelPath: modelDir.path)
        
        if voskModel?.model == nil {
            throw NSError(domain: "SpeechToTextPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to load Vosk model from: \(modelDir.path)"])
        }
        
        print("Vosk model loaded successfully")
    }
    
    private func startVoskRecognition(language: String) {
        guard let voskModel = voskModel else {
            print("ERROR: Vosk model not initialized")
            return
        }
        
        print("Starting Vosk recognition...")
        
        do {
            // Configure audio session
            let audioSession = AVAudioSession.sharedInstance()
            try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
            try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
            print("Audio session configured successfully")
            
            // Create a new audio engine
            audioEngine = AVAudioEngine()
            guard let audioEngine = audioEngine else {
                print("ERROR: Unable to create audio engine")
                return
            }
            
            let inputNode = audioEngine.inputNode
            let formatInput = inputNode.inputFormat(forBus: 0)
            print("Input format: sampleRate=\(formatInput.sampleRate), channels=\(formatInput.channelCount)")
            
            let formatPcm = AVAudioFormat(commonFormat: AVAudioCommonFormat.pcmFormatInt16, 
                                        sampleRate: formatInput.sampleRate, 
                                        channels: 1, 
                                        interleaved: true)
            
            guard let formatPcm = formatPcm else {
                print("ERROR: Unable to create PCM format")
                return
            }
            
            // Create Vosk recognizer
            print("Creating Vosk recognizer with sample rate: \(formatInput.sampleRate)")
            voskRecognizer = VoskRecognizer(model: voskModel, sampleRate: Float(formatInput.sampleRate))
            guard let voskRecognizer = voskRecognizer else {
                print("ERROR: Unable to create Vosk recognizer")
                return
            }
            print("Vosk recognizer created successfully")
            
            inputNode.installTap(onBus: 0,
                                 bufferSize: UInt32(formatInput.sampleRate / 10),
                                 format: formatPcm) { buffer, time in
                                    self.processingQueue?.async {
                                        let result = voskRecognizer.recognizeData(buffer: buffer)
                                        
                                        // Parse JSON result
                                        if let data = result.data(using: .utf8) {
                                            do {
                                                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                                                    let text = json["text"] as? String ?? ""
                                                    let partial = json["partial"] as? String ?? ""
                                                    
                                                    DispatchQueue.main.async {
                                                        // Notify listeners with recognition result
                                                        if !text.isEmpty {
                                                            print("Final result: \(text)")
                                                            self.notifyListeners("recognitionResult", data: [
                                                                "text": text,
                                                                "isFinal": true,
                                                                "language": language
                                                            ])
                                                        } else if !partial.isEmpty {
                                                            print("Partial result: \(partial)")
                                                            self.notifyListeners("recognitionResult", data: [
                                                                "text": partial,
                                                                "isFinal": false,
                                                                "language": language
                                                            ])
                                                        }
                                                    }
                                                }
                                            } catch {
                                                print("Error parsing Vosk result: \(error)")
                                            }
                                        }
                                    }
            }
            
            // Start the stream of audio data
            audioEngine.prepare()
            try audioEngine.start()
            isRecording = true
            currentLanguage = language
            print("Audio engine started successfully")
            
        } catch {
            print("ERROR: Unable to start AVAudioEngine: \(error.localizedDescription)")
        }
    }
    
    private func stopVoskRecognition() {
        if let audioEngine = audioEngine {
            audioEngine.stop()
            audioEngine.inputNode.removeTap(onBus: 0)
        }
        
        voskRecognizer = nil
        isRecording = false
        
        // Deactivate audio session
        do {
            try AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        } catch {
            print("Audio session deactivation failed: \(error)")
        }
    }
    
    deinit {
        stopVoskRecognition()
    }
}

File: android/src/main/java/dev/ravi/offlinespeech/OfflineSpeechRecognitionPlugin.java
package dev.ravi.offlinespeech;

import org.vosk.Model;
import org.vosk.Recognizer;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import java.io.File;
import java.util.Map;
import java.util.HashMap;
import java.io.IOException;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;

import android.Manifest;
import android.util.Log;
import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.content.pm.PackageManager;

import androidx.core.app.ActivityCompat;
import androidx.annotation.RequiresPermission;

import com.getcapacitor.Plugin;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(
    name = "OfflineSpeechRecognition",
    permissions = {
        @Permission(
            strings = {
                Manifest.permission.INTERNET,
                Manifest.permission.RECORD_AUDIO,
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            },
            alias = "audio"
        )
    }
)
public class OfflineSpeechRecognitionPlugin extends Plugin {

    private static final String TAG = "OfflineSpeechRecognitionPlugin";

    private Model currentModel;
    private Recognizer recognizer;
    private AudioRecord audioRecord;
    private boolean isRecording = false;
    private ExecutorService executorService;
    private String currentLanguage = "en-us";
    private ModelDownloadManager modelDownloadManager;

    // supported languages mapping to model names
    private static final Map<String, String> SUPPORTED_LANGUAGES = new HashMap<String, String>() {{
        put("en-us", "model-en");
        put("de", "model-de");
        put("fr", "model-fr");
        put("es", "model-es");
        put("pt", "model-pt");
        put("zh", "model-zh");
        put("ru", "model-ru");
        put("tr", "model-tr");
        put("vi", "model-vi");
        put("it", "model-it");
        put("hi", "model-hi");
        put("gu", "model-gu");
        put("te", "model-te");
        put("ja", "model-ja");
        put("ko", "model-ko");
    }};

    @Override
    public void load() {
        super.load();
        executorService = Executors.newCachedThreadPool();
        modelDownloadManager = new ModelDownloadManager(getContext());
    }

    @PluginMethod()
    public void echo(PluginCall call) {
        String value = call.getString("value");
        call.resolve(new JSObject().put("value", value));
    }

    @PluginMethod()
    public void getSupportedLanguages(PluginCall call) {
        try {
            JSONArray languages = new JSONArray();
            for (Map.Entry<String, String> entry : SUPPORTED_LANGUAGES.entrySet()) {
                JSONObject language = new JSONObject();

                language.put("code", entry.getKey());
                language.put("modelName", entry.getValue());
                language.put("name", getLanguageName(entry.getKey()));

                languages.put(language);
            }

            JSObject result = new JSObject();
            result.put("languages", languages);
            call.resolve(result);
        } catch (JSONException e) {
            Log.e(TAG, "Error creating supported languages response", e);
            call.reject("Error getting supported languages", e);
        }
    }

    @PluginMethod()
    public void getDownloadedLanguageModels(PluginCall call) {
        try {
            JSONArray downloadedModels = new JSONArray();

            // check each supported language to see if model is downloaded
            for (Map.Entry<String, String> entry : SUPPORTED_LANGUAGES.entrySet()) {
                String modelName = entry.getValue();
                String languageCode = entry.getKey();

                if (modelDownloadManager.isModelDownloaded(modelName)) {
                    JSONObject model = new JSONObject();

                    model.put("modelName", modelName);
                    model.put("language", languageCode);
                    model.put("name", getLanguageName(languageCode));
                    model.put("size", modelDownloadManager.getModelSize(modelName));

                    downloadedModels.put(model);
                }
            }

            JSObject result = new JSObject();
            result.put("models", downloadedModels);
            call.resolve(result);
        } catch (JSONException e) {
            Log.e(TAG, "Error getting downloaded models", e);
            call.reject("Error getting downloaded models", e);
        }
    }

    @PluginMethod()
    public void downloadLanguageModel(PluginCall call) {
        String language = call.getString("language");
        if (language == null || !SUPPORTED_LANGUAGES.containsKey(language)) {
            call.reject("Invalid or unsupported language code");
            return;
        }

        String modelName = SUPPORTED_LANGUAGES.get(language);

        // check if model is already downloaded
        if (modelDownloadManager.isModelDownloaded(modelName)) {
            JSObject result = new JSObject();

            result.put("success", true);
            result.put("language", language);
            result.put("message", "Model already downloaded");

            call.resolve(result);
            return;
        }

        // check if download is already in progress
        if (modelDownloadManager.isDownloadInProgress(modelName)) {
            call.reject("Download already in progress for this model");
            return;
        }

        if (!hasAudioPermissions()) {
            requestAudioPermissions(call);
            return;
        }

        // use ModelDownloadManager to download the model
        modelDownloadManager.downloadModel(
            modelName,
            progress -> {
                JSObject progressData = new JSObject();
                progressData.put("progress", progress);
                progressData.put("message", "Downloading model... " + progress + "%");
                notifyListeners("downloadProgress", progressData);
            },
            () -> {
                JSObject result = new JSObject();
                result.put("success", true);
                result.put("language", language);
                result.put("modelName", modelName);
                call.resolve(result);
            },
            error -> call.reject("Error downloading model: " + error)
        );
    }

    @PluginMethod()
    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
    public void startRecognition(PluginCall call) {
        String language = call.getString("language", "en-us");

        if (!hasAudioPermissions()) {
            requestAudioPermissions(call);
            return;
        }

        String modelName = SUPPORTED_LANGUAGES.get(language);
        if (modelName == null) {
            call.reject("Unsupported language: " + language);
            return;
        }

        if (!modelDownloadManager.isModelDownloaded(modelName)) {
            call.reject("Language model not downloaded. Please download the model first.");
            return;
        }

        try {
            loadModel(modelName);
            startAudioRecording();
            call.resolve();
        } catch (Exception e) {
            Log.e(TAG, "Error starting recognition", e);
            call.reject("Error starting recognition: " + e.getMessage());
        }
    }

    @PluginMethod()
    public void stopRecognition(PluginCall call) {
        try {
            stopAudioRecording();
            if (recognizer != null) {
                recognizer.close();
                recognizer = null;
            }
            call.resolve();
        } catch (Exception e) {
            Log.e(TAG, "Error stopping recognition", e);
            call.reject("Error stopping recognition: " + e.getMessage());
        }
    }

    @PermissionCallback
    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
    private void audioPermissionsCallback(PluginCall call) {
        if (hasAudioPermissions()) {
            String method = call.getMethodName();
            switch (method) {
                case "downloadLanguageModel":
                    downloadLanguageModel(call);
                    break;
                case "startRecognition":
                    startRecognition(call);
                    break;
            }
        } else {
            call.reject("Audio permissions are required for speech recognition");
        }
    }

    private boolean hasAudioPermissions() {
        return ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestAudioPermissions(PluginCall call) {
        requestPermissionForAlias("audio", call, "audioPermissionsCallback");
    }

    private void loadModel(String modelName) throws IOException {
        File modelDir = modelDownloadManager.getModelDirectory(modelName);

        if (!modelDir.exists()) {
            throw new IOException("Model directory not found: " + modelDir.getAbsolutePath());
        }

        if (!modelDownloadManager.verifyModel(modelDir)) {
            throw new IOException("Model verification failed for: " + modelDir.getAbsolutePath() + ". Model files may be corrupted or incomplete.");
        }

        Log.d(TAG, "Loading Vosk model from: " + modelDir.getAbsolutePath());

        try {
            currentModel = new Model(modelDir.getAbsolutePath());
            recognizer = new Recognizer(currentModel, 16000.0f);
            Log.d(TAG, "Vosk model loaded successfully");
        } catch (Exception e) {
            Log.e(TAG, "Failed to load Vosk model from: " + modelDir.getAbsolutePath(), e);
            throw new IOException("Failed to load Vosk model: " + e.getMessage(), e);
        }

        // find the language code for this model name
        for (Map.Entry<String, String> entry : SUPPORTED_LANGUAGES.entrySet()) {
            if (entry.getValue().equals(modelName)) {
                currentLanguage = entry.getKey();
                break;
            }
        }
    }

    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
    private void startAudioRecording() {
        int sampleRate = 16000;
        int channelConfig = AudioFormat.CHANNEL_IN_MONO;
        int audioFormat = AudioFormat.ENCODING_PCM_16BIT;

        int bufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat);
        audioRecord = new AudioRecord(MediaRecorder.AudioSource.MIC, sampleRate, channelConfig, audioFormat, bufferSize);

        audioRecord.startRecording();
        isRecording = true;

        // start recognition thread
        executorService.execute(this::processAudioData);
    }

    private void stopAudioRecording() {
        isRecording = false;
        if (audioRecord != null) {
            audioRecord.stop();
            audioRecord.release();
            audioRecord = null;
        }
    }

    private void processAudioData() {
        byte[] buffer = new byte[4096];

        while (isRecording && audioRecord != null) {
            int bytesRead = audioRecord.read(buffer, 0, buffer.length);
            if (bytesRead > 0 && recognizer != null) {
                if (recognizer.acceptWaveForm(buffer, bytesRead)) {
                    // final result
                    String result = recognizer.getResult();
                    notifyRecognitionResult(result, true);
                } else {
                    // partial result
                    String partialResult = recognizer.getPartialResult();
                    notifyRecognitionResult(partialResult, false);
                }
            }
        }

        // get final result
        if (recognizer != null) {
            String finalResult = recognizer.getFinalResult();
            notifyRecognitionResult(finalResult, true);
        }
    }

    private void notifyRecognitionResult(String result, boolean isFinal) {
        try {
            JSONObject resultJson = new JSONObject(result);
            String text = resultJson.optString("text", "");

            if (!text.isEmpty()) {
                JSObject recognitionData = new JSObject();
                recognitionData.put("text", text);
                recognitionData.put("isFinal", isFinal);
                recognitionData.put("language", currentLanguage);
                notifyListeners("recognitionResult", recognitionData);
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing recognition result", e);
        }
    }

    private String getLanguageName(String code) {
        Map<String, String> languageNames = new HashMap<String, String>() {{
            put("hi", "Hindi");
            put("ko", "Korean");
            put("de", "German");
            put("fr", "French");
            put("te", "Telugu");
            put("es", "Spanish");
            put("zh", "Chinese");
            put("ru", "Russian");
            put("tr", "Turkish");
            put("it", "Italian");
            put("gu", "Gujarati");
            put("ja", "Japanese");
            put("pt", "Portuguese");
            put("vi", "Vietnamese");
            put("en-us", "English (US)");
            put("en-in", "English (India)");
        }};

        return languageNames.getOrDefault(code, code);
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();

        if (isRecording) {
            stopAudioRecording();
        }

        if (recognizer != null) {
            recognizer.close();
            recognizer = null;
        }

        if (currentModel != null) {
            currentModel.close();
            currentModel = null;
        }

        if (executorService != null) {
            executorService.shutdown();
            executorService = null;
        }
    }
}
